Model/Service/Index.php

```diff
@@ -0,0 +1,435 @@
+     * @throws NostoException
+     */
+    public function invalidateOrCreateProductOrParent(Product $product, Store $store)
+    {
+        $parents = $this->nostoProductRepository->resolveParentProductIds($product);
+
+        //Products has no parents and Index product itself
+        if (empty($parents)) {
+            $this->updateOrCreateDirtyEntity($product, $store);
+            $this->invalidatedProducts[] = $product->getId();
+            return;
+        }
+
+        //Loop through product parents and Index the parents
+        if (is_array($parents)) {
+            $this->invalidateOrCreateParents($parents, $store);
+            return;
+        }
+
+        throw new NostoException('Could not index product with id: ' . $product->getId());
+    }
+
+    /**
+     * @param array $ids
+     * @param Store $store
+     * @throws NostoException
+     */
+    private function invalidateOrCreateParents(array $ids, Store $store)
+    {
+        $collection = $this->productCollectionFactory->create();
+        $collection->addIdsToFilter($ids);
+        $collection->setPageSize(self::PRODUCT_DATA_BATCH_SIZE);
+        $iterator = new Iterator($collection);
+        $iterator->each(function (Product $product) use ($store) {
+            if ($this->hasParentBeenInvalidated($product->getId()) === false) {
+                $this->updateOrCreateDirtyEntity($product, $store);
+                $this->invalidatedProducts[] = $product->getId();
+            }
+        });
+    }
+
+    /**
+     * @param $productId
+     * @return bool
+     */
+    private function hasParentBeenInvalidated($productId)
+    {
+        if (in_array($productId, $this->invalidatedProducts, false)) {
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * @param ProductInterface $product
+     * @param StoreInterface $store
+     * @return void
+     */
+    public function updateOrCreateDirtyEntity(ProductInterface $product, StoreInterface $store)
+    {
+        $indexedProduct = $this->indexRepository->getByProductIdAndStoreId($product->getId(), $store->getId());
+        try {
+            if ($indexedProduct === null) { // Creates Index Product
+                $indexedProduct = $this->indexBuilder->build($product, $store);
+            }
+            $indexedProduct->setIsDirty(true);
+            $indexedProduct->setUpdatedAt($this->magentoTimeZone->date());
+            $this->indexRepository->save($indexedProduct);
+        } catch (Exception $e) {
+            $this->getLogger()->exception($e);
+        }
+    }
+
+    /**
+     * @param Store $store
+     * @param array $ids
+     * @throws NostoException
+     */
+    public function indexProducts(Store $store, array $ids = [])
+    {
+        $account = $this->nostoHelperAccount->findAccount($store);
+        if ($account instanceof NostoSignupAccount === false) {
+            throw new NostoException(sprintf('Store view %s does not have Nosto installed', $store->getName()));
```

Comment:
The other indexer checks if the account is connected. This one does it differently. I would change both indexers to do it the same way in the same method.